### PR TITLE
kubelet: Refactor kubelet.runContainer.

### PR DESCRIPTION
**Pull request by yifan-gu**
This PR tries to move the launch container logic from kubelet to container runtime. This is a prelude for fixing #5896 

The next steps will be:
1: Move the `killContainer` logic to container runtime (will need to pull the readiness manager out, so container runtime can update the readiness after killing a container).

2: Wraps the these functions with `RunContainerInPod`, in which the container runtime will handle the necessary restart of the pod infra container and the whole pod.

3: After 1) and 2), I think we are close to simplify the syncPod's logic into a more generic one.
Basically, it will be something like:

```
for each container in Spec:
    if it's found in runningPod:
        if container hash changed || container is unhealthy
            restart the container
    else:
        if needs to restart:
            restart the container

for each unidentified containers in the runningPod:
    kill container

```

### CODE REVIEW

**Comment by yifan-gu** [comment]
The PR is WIP because I need to introduce reference manager first, so the docker container runtime can set reference after creating the container.

Happy to hear your opinions :)
@dchen1107 @vmarmol

File: pkg/kubelet/dockertools/docker.go [hunk 1]
```diff
@@ -46,6 +48,13 @@ const (
 	DockerPrefix          = "docker://"
 )
 
+const (
```
Comment (xiang90) [inline pkg/kubelet/dockertools/docker.go]: will rocket also needs this? does it make sense to move this to container pkg?

Comment (yifan-gu) [inline pkg/kubelet/dockertools/docker.go]: @jonboulle 
Need clarification about how we represents CPU share in rocket?

File: pkg/kubelet/dockertools/docker.go [hunk 2]
```diff
@@ -866,3 +875,141 @@ func GetPods(client DockerInterface, all bool) ([]*kubecontainer.Pod, error) {
...
+	}
+	// Conceptually (milliCPU / milliCPUToCPU) * sharesPerCPU, but factored to improve rounding.
+	shares := (milliCPU * sharesPerCPU) / milliCPUToCPU
+	if shares < minShares {
+		return minShares
+	}
+	return shares
+}
+
+func makePortsAndBindings(container *api.Container) (map[docker.Port]struct{}, map[docker.Port][]docker.PortBinding) {
```
Comment (xiang90) [inline pkg/kubelet/dockertools/docker.go]: this seems to be a generic thing too.

File: pkg/kubelet/dockertools/docker.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/dockertools/docker.go]: @xiang90 Just checked spec. Think we just use milliCore.

File: pkg/kubelet/dockertools/docker.go (hunk 2 above)
Comment (yifan-gu) [inline pkg/kubelet/dockertools/docker.go]: @xiang90 Hmm, I think in the future after we land this https://github.com/coreos/rocket/issues/294 then we can generate podmanifest, so we don't need these helpers for rocket.

File: pkg/kubelet/dockertools/docker.go (hunk 1 above)
Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: FWIU Rocket will be doing a similar conversion internally. Docker exposes shares so we do the conversion ourselves. The API we'd like to expose the Kubelet is millicores.

File: pkg/kubelet/container/runtime.go [hunk 3]
```diff
@@ -89,6 +89,27 @@ type Container struct {
...
+	// 'hostPath:containerPath', or
+	// 'hostPath:containerPath:ro', if the path read only.
+	Binds []string
+	// If the container has specified the TerminationMessagePath, then
+	// this path will be used to mount to that path.
+	ContainerLogPath string
+	// The DNS settings for the container.
+	DNS       []string
+	DNSSearch []string
+	// Namespaces for the container.
```
Comment (vmarmol) [inline pkg/kubelet/container/runtime.go]: "Docker namespace identifiers"

Comment (vmarmol) [inline pkg/kubelet/container/runtime.go]: Ah this is for the runtime, we'll need more description here since its not clear. Which does Rocket support

File: pkg/kubelet/container/runtime.go [hunk 4]
```diff
@@ -89,6 +89,27 @@ type Container struct {
...
+	// The environment variables, they are in the form of 'key=value'.
+	Envs []string
+	// The mounts for the containers, they are in the form of:
+	// 'hostPath:containerPath', or
+	// 'hostPath:containerPath:ro', if the path read only.
+	Binds []string
+	// If the container has specified the TerminationMessagePath, then
+	// this path will be used to mount to that path.
+	ContainerLogPath string
+	// The DNS settings for the container.
```
Comment (vmarmol) [inline pkg/kubelet/container/runtime.go]: Can we  document what these fields are.

File: pkg/kubelet/dockertools/docker.go [hunk 5]
```diff
@@ -866,3 +875,141 @@ func GetPods(client DockerInterface, all bool) ([]*kubecontainer.Pod, error) {
...
+	for _, cap := range capAdd {
+		addCaps = append(addCaps, string(cap))
+	}
+	for _, cap := range capDrop {
+		dropCaps = append(dropCaps, string(cap))
+	}
+	return addCaps, dropCaps
+}
+
+func RunContainer(client DockerInterface, container api.Container, pod *api.Pod, ref *api.ObjectReference, recorder record.EventRecorder, opts *kubecontainer.RunContainerOptions) (string, error) {
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: Should we add a TODO to unify this with the container runtime API?

File: pkg/kubelet/dockertools/docker.go [hunk 6]
```diff
@@ -866,3 +875,141 @@ func GetPods(client DockerInterface, all bool) ([]*kubecontainer.Pod, error) {
...
+	dockerContainer, err := client.CreateContainer(dockerOpts)
+	if err != nil {
+		if ref != nil {
+			recorder.Eventf(ref, "failed", "Failed to create docker container with error: %v", err)
+		}
+		return "", err
+	}
+	// Remember this reference so we can report events about this container
+	if ref != nil {
+		kl.setRef(dockerContainer.ID, ref)
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: I'm guessing these references are waiting on the other PR :)

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -794,48 +695,43 @@ func (kl *Kubelet) runContainer(pod *api.Pod, container *api.Container, podVolum
...
-		if ref != nil {
-			kl.recorder.Eventf(ref, "failed",
-				"Failed to start with docker id %v with error: %v", dockerContainer.ID, err)
-		}
 		return "", err
 	}
-	if ref != nil {
-		kl.recorder.Eventf(ref, "started", "Started with docker id %v", dockerContainer.ID)
+
+	// TODO(yifan): Replace with RunContainerInPod, so we can elimiate 'netMode', 'ipcMode'
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: typo: eliminate

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -927,27 +823,29 @@ func (kl *Kubelet) makeEnvironmentVariables(ns string, container *api.Container)
 	return result, nil
 }
 
-func (kl *Kubelet) applyClusterDNS(hc *docker.HostConfig, pod *api.Pod) error {
+func (kl *Kubelet) getClusterDNS(pod *api.Pod) ([]string, []string, error) {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: nit: can we document the outputs :)

File: pkg/kubelet/dockertools/docker.go (hunk 5 above)
Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: Ah I see you already have the TODO in kubelet.go.

**Comment by vmarmol** [comment]
Great refactoring @yifan-gu! Just some comments.

File: pkg/kubelet/dockertools/docker.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/dockertools/docker.go]: @vmarmol Awesome! Thanks :)

**Comment by yifan-gu** [comment]
@vmarmol Ready for another review :)
Updated: Move the creating/mouting termination log file logic into the container runtime (because otherwise we cannot get the id of the container before actually starting it). Had added comments for that.

PTAL THX :)
